concert/cloudformation.choice-aggregator.yml

```diff
@@ -0,0 +1,282 @@
+          - Effect: Allow
+            Principal:
+              Service: kinesisanalytics.amazonaws.com
+            Action: sts:AssumeRole
+      Policies:
+        - PolicyName: !Sub ${AWS::Region}-${Environment}-enchanted-brain-kinesis-analytics
+          PolicyDocument:
+            Version: 2012-10-17
+            Statement:
+              - Effect: Allow
+                Action:
+                  - firehose:DescribeDeliveryStream
+                  - firehose:Get*
+                Resource: !GetAtt KinesisFirehoseStream.Arn
+              - Effect: Allow
+                Action: lambda:InvokeFunction
+                Resource: !GetAtt AggregateWriterFunction.Arn
+              - Effect: Allow
+                Action: logs:PutLogEvents
+                Resource: !Sub arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:/aws/kinesisfirehose/${Environment}-enchanted-brain-kinesis:log-stream:${KinesisAnalyticsLogStream}
+
+  KinesisAnalyticsApplication:
+    Type: AWS::KinesisAnalyticsV2::Application
+    Properties:
+      ApplicationName: !Sub ${Environment}-enchanted-brain-aggregate
+      RuntimeEnvironment: SQL-1_0
+      ServiceExecutionRole: !GetAtt KinesisAnalyticsApplicationRole.Arn
+      ApplicationConfiguration:
+        ApplicationCodeConfiguration:
+          CodeContent:
+            TextContent: !Sub
+              - |
+                CREATE OR REPLACE STREAM ${DestinationStream}
+                (
+                  CHOICE_TIME   TIMESTAMP,
+                  CHOICE_TYPE   VARCHAR(32),
+                  CHOICE_SUM    DOUBLE,
+                  CHOICE_COUNT  int
+                );
+                CREATE OR REPLACE PUMP SUM_PUMP AS
+                  INSERT INTO ${DestinationStream}
+                  SELECT STREAM CHOICE_TIME,
+                    CHOICE_TYPE,
+                    SUM(CAST(CHOICE AS DOUBLE)) OVER (
+                      PARTITION BY CHOICE_TYPE
+                      RANGE INTERVAL '1' MINUTE PRECEDING
+                    ) AS choiceSum,
+                    COUNT(*) OVER W1 AS choiceCount
+                  FROM ${SourceStreamPrefix}_001 AS SOURCE
+                  WHERE CHOICE_TYPE NOT LIKE 'CHOICE_COLOR'
+                  WINDOW W1 AS (
+                    PARTITION BY CHOICE_TIME
+                    RANGE INTERVAL '1' MINUTE PRECEDING
+                  );
+                CREATE OR REPLACE PUMP COUNT_PUMP AS
+                  INSERT INTO ${DestinationStream}
+                  SELECT STREAM CHOICE_TIME,
+                    CHOICE_TYPE || '_' || CHOICE,
+                    COUNT(CHOICE) OVER (
+                      PARTITION BY CHOICE_TYPE, CHOICE
+                      RANGE INTERVAL '1' MINUTE PRECEDING
+                    ) AS choiceSum,
+                    COUNT(*) OVER W1 AS choiceCount
+                  FROM ${SourceStreamPrefix}_001 AS SOURCE
+                  WHERE CHOICE_TYPE LIKE 'CHOICE_COLOR'
+                  WINDOW W1 AS (
+                    PARTITION BY CHOICE_TIME
```

Comment:
I think you also need to partition by `CHOICE` for color. Then you can simply choiceSum and choiceCount to just be `COUNT(*) OVER W1`

Reply:
Yep, this is what I ended up doing for the most recent revision after our discussion. At this point, I was still trying to figure out how to set choiceCount to the number of choices for all colors at this timestamp. Out of curiosity, do you know what would have been the right way to make that happen?

Reply:
Two different windows, one partitioning by CHOICE and one not